File: python/src/spicedb_embedded/embedded.py
"""Embedded SpiceDB instance.

Uses in-memory transport: unary RPCs via FFI, streaming APIs (Watch, ReadRelationships, etc.)
via a streaming proxy. Use permissions(), schema(), and watch() to access the full SpiceDB API.
"""

import grpc
from authzed.api.v1 import (
    CheckBulkPermissionsRequest,
    CheckBulkPermissionsResponse,
    CheckPermissionRequest,
    CheckPermissionResponse,
    DeleteRelationshipsRequest,
    DeleteRelationshipsResponse,
    ExpandPermissionTreeRequest,
    ExpandPermissionTreeResponse,
    ReadRelationshipsRequest,
    ReadSchemaRequest,
    ReadSchemaResponse,
    Relationship,
    RelationshipUpdate,
    WriteRelationshipsRequest,
    WriteRelationshipsResponse,
    WriteSchemaRequest,
    WriteSchemaResponse,
)
from authzed.api.v1.permission_service_pb2_grpc import PermissionsServiceStub
from authzed.api.v1.watch_service_pb2_grpc import WatchServiceStub

from spicedb_embedded.errors import SpiceDBError
from spicedb_embedded.ffi import (
    spicedb_dispose,
    spicedb_permissions_check_bulk_permissions,
    spicedb_permissions_check_permission,
    spicedb_permissions_delete_relationships,
    spicedb_permissions_expand_permission_tree,
    spicedb_permissions_write_relationships,
    spicedb_schema_read_schema,
    spicedb_schema_write_schema,
    spicedb_start,
)
from spicedb_embedded.tcp_channel import get_target as get_tcp_target
from spicedb_embedded.unix_socket_channel import get_target as get_unix_socket_target


def _streaming_target(streaming_address: str, streaming_transport: str) -> str:
    """Return gRPC target for streaming_address using streaming_transport ('unix' or 'tcp')."""
    if streaming_transport == "unix":
        return get_unix_socket_target(streaming_address)
    return get_tcp_target(streaming_address)


class EmbeddedSpiceDB:
    """
    Embedded SpiceDB instance (in-memory only).

    Unary RPCs go through FFI; streaming (Watch, ReadRelationships, etc.) goes through
    the streaming proxy. Use permissions(), schema(), and watch() to access the API.
    """

    def __init__(
        self,
        schema: str,
        relationships: list[Relationship] | None = None,
        *,
        options: dict | None = None,
    ):
        """
        Create an embedded SpiceDB instance.

        Args:
            schema: The SpiceDB schema definition (ZED language).
            relationships: Initial relationships. Defaults to [].
            options: Optional config (datastore, datastore_uri, etc.). Instance is always in-memory.
        """
        data = spicedb_start(options)
        self._handle = data["handle"]
        self._streaming_address = data["streaming_address"]
        self._streaming_transport = data["streaming_transport"]
        target = _streaming_target(self._streaming_address, self._streaming_transport)
        self._channel = grpc.insecure_channel(target)

        try:
            self._bootstrap(schema, relationships or [])
        except Exception as e:
            self.close()
            raise SpiceDBError(f"Failed to bootstrap: {e}") from e

    def _bootstrap(self, schema: str, relationships: list[Relationship]) -> None:
        write_schema_req = WriteSchemaRequest(schema=schema)
        spicedb_schema_write_schema(self._handle, write_schema_req.SerializeToString())

        if relationships:
            updates = [
                RelationshipUpdate(
                    operation=RelationshipUpdate.Operation.OPERATION_TOUCH,
                    relationship=r,
                )
                for r in relationships
            ]
            write_rel_req = WriteRelationshipsRequest(updates=updates)
            spicedb_permissions_write_relationships(
                self._handle, write_rel_req.SerializeToString()
            )

    def permissions(self) -> "EmbeddedPermissionsStub":
        """Permissions service: unary via FFI, streaming (ReadRelationships, etc.) via proxy."""
        return EmbeddedPermissionsStub(self._handle, self._channel)

    def schema(self) -> "EmbeddedSchemaStub":
        """Schema service (ReadSchema, WriteSchema) via FFI."""
        return EmbeddedSchemaStub(self._handle)

    def watch(self) -> WatchServiceStub:
        """Watch service (streaming) via proxy."""
        return WatchServiceStub(self._channel)

    def channel(self) -> grpc.Channel:
        """Channel connected to the streaming proxy."""
        return self._channel

    def streaming_address(self) -> str:
        """Streaming proxy address (Unix path or host:port)."""
        return self._streaming_address

    def streaming_transport(self) -> str:
        """Streaming proxy transport: 'unix' or 'tcp'."""
        return self._streaming_transport

    def close(self) -> None:
        """Dispose the instance and close the channel."""
        spicedb_dispose(self._handle)
        self._channel.close()

    def __enter__(self) -> "EmbeddedSpiceDB":
        return self

    def __exit__(self, *args) -> None:
        self.close()


class EmbeddedPermissionsStub:
    """Permissions client: unary via FFI, ReadRelationships via streaming channel."""

    def __init__(self, handle: int, channel: grpc.Channel) -> None:
        self._handle = handle
        self._stub = PermissionsServiceStub(channel)

    def CheckPermission(
        self, request: CheckPermissionRequest
    ) -> CheckPermissionResponse:
        raw = spicedb_permissions_check_permission(
            self._handle, request.SerializeToString()
        )
        r = CheckPermissionResponse()
        r.ParseFromString(raw)
        return r

    def WriteRelationships(
        self, request: WriteRelationshipsRequest
    ) -> WriteRelationshipsResponse:
        raw = spicedb_permissions_write_relationships(
            self._handle, request.SerializeToString()
        )
        r = WriteRelationshipsResponse()
        r.ParseFromString(raw)
        return r

    def DeleteRelationships(
        self, request: DeleteRelationshipsRequest
    ) -> DeleteRelationshipsResponse:
        raw = spicedb_permissions_delete_relationships(
            self._handle, request.SerializeToString()
        )
        r = DeleteRelationshipsResponse()
        r.ParseFromString(raw)
        return r

    def CheckBulkPermissions(
        self, request: CheckBulkPermissionsRequest
    ) -> CheckBulkPermissionsResponse:
        raw = spicedb_permissions_check_bulk_permissions(
            self._handle, request.SerializeToString()
        )
        r = CheckBulkPermissionsResponse()
        r.ParseFromString(raw)
        return r

    def ExpandPermissionTree(
        self, request: ExpandPermissionTreeRequest
    ) -> ExpandPermissionTreeResponse:
        raw = spicedb_permissions_expand_permission_tree(
            self._handle, request.SerializeToString()
        )
        r = ExpandPermissionTreeResponse()
        r.ParseFromString(raw)
        return r

    def ReadRelationships(self, request: ReadRelationshipsRequest, **kwargs):
        """Streaming: uses the streaming proxy."""
        return self._stub.ReadRelationships(request, **kwargs)


class EmbeddedSchemaStub:
    """Schema client (ReadSchema, WriteSchema) via FFI."""

    def __init__(self, handle: int) -> None:
        self._handle = handle

    def ReadSchema(self, request: ReadSchemaRequest) -> ReadSchemaResponse:
        raw = spicedb_schema_read_schema(self._handle, request.SerializeToString())
        r = ReadSchemaResponse()
        r.ParseFromString(raw)
        return r

    def WriteSchema(self, request: WriteSchemaRequest) -> WriteSchemaResponse:
        raw = spicedb_schema_write_schema(self._handle, request.SerializeToString())
        r = WriteSchemaResponse()
        r.ParseFromString(raw)
        return r

File: node/src/embedded.ts
/**
 * Embedded SpiceDB instance.
 *
 * Uses the C-shared library with in-memory transport: unary RPCs go through FFI,
 * streaming APIs (Watch, ReadRelationships, etc.) use a streaming proxy.
 *
 * Use permissions(), schema(), and watch() to access the full SpiceDB API.
 */

import * as grpc from "@grpc/grpc-js";
import { v1 } from "@authzed/authzed-node";
import {
  spicedb_dispose,
  spicedb_schema_read_schema,
  spicedb_schema_write_schema,
  spicedb_permissions_check_permission,
  spicedb_permissions_write_relationships,
  spicedb_permissions_delete_relationships,
  spicedb_permissions_check_bulk_permissions,
  spicedb_permissions_expand_permission_tree,
  spicedb_start,
  type SpiceDBStartOptions,
} from "./ffi.js";
import { getTarget as getTcpTarget } from "./tcp_channel.js";
import { getTarget as getUnixSocketTarget } from "./unix_socket_channel.js";

const {
  CheckPermissionRequest,
  CheckPermissionResponse,
  CheckBulkPermissionsRequest,
  CheckBulkPermissionsResponse,
  ExpandPermissionTreeRequest,
  ExpandPermissionTreeResponse,
  DeleteRelationshipsRequest,
  DeleteRelationshipsResponse,
  WriteRelationshipsRequest,
  WriteRelationshipsResponse,
  ReadSchemaRequest,
  ReadSchemaResponse,
  WriteSchemaRequest,
  WriteSchemaResponse,
  NewClientWithChannelCredentials,
  RelationshipUpdate,
  RelationshipUpdate_Operation,
} = v1;

export class SpiceDBError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SpiceDBError";
  }
}

/** Build gRPC target from streaming_address (Unix path or host:port). */
function streamingTarget(streamingAddress: string): string {
  return streamingAddress.startsWith("/")
    ? getUnixSocketTarget(streamingAddress)
    : getTcpTarget(streamingAddress);
}

export class EmbeddedSpiceDB {
  private readonly handle: number;
  private readonly _streamingAddress: string;
  private readonly streamingClient: v1.ZedClientInterface;
  private _closed = false;

  private constructor(
    handle: number,
    streamingAddr: string,
    streamingClient: v1.ZedClientInterface
  ) {
    this.handle = handle;
    this._streamingAddress = streamingAddr;
    this.streamingClient = streamingClient;
  }

  /**
   * Create a new embedded SpiceDB instance (in-memory only).
   *
   * @param schema - The SpiceDB schema definition (ZED language)
   * @param relationships - Initial relationships (empty array allowed)
   * @param options - Optional datastore options (no grpc_transport; always memory)
   * @returns New EmbeddedSpiceDB instance
   */
  static async create(
    schema: string,
    relationships: v1.Relationship[] = [],
    options?: SpiceDBStartOptions | null
  ): Promise<EmbeddedSpiceDB> {
    const data = spicedb_start(options ?? undefined);
    const { handle, streaming_address } = data;

    const target = streamingTarget(streaming_address);
    const creds = grpc.credentials.createInsecure();
    const streamingClient = NewClientWithChannelCredentials(target, creds, 0);

    const db = new EmbeddedSpiceDB(handle, streaming_address, streamingClient);

    try {
      await db.bootstrap(schema, relationships);
    } catch (e) {
      db.close();
      throw new SpiceDBError(
        `Failed to bootstrap: ${e instanceof Error ? e.message : String(e)}`
      );
    }

    return db;
  }

  private async bootstrap(
    schema: string,
    relationships: v1.Relationship[]
  ): Promise<void> {
    const schemaRequest = WriteSchemaRequest.create({ schema });
    const schemaBytes = new Uint8Array(
      WriteSchemaRequest.toBinary(schemaRequest) as ArrayLike<number>
    );
    spicedb_schema_write_schema(this.handle, schemaBytes);

    if (relationships.length > 0) {
      const updates = relationships.map((r) =>
        RelationshipUpdate.create({
          operation: RelationshipUpdate_Operation.TOUCH,
          relationship: r,
        })
      );
      const writeRequest = WriteRelationshipsRequest.create({ updates });
      const writeBytes = new Uint8Array(
        WriteRelationshipsRequest.toBinary(writeRequest) as ArrayLike<number>
      );
      spicedb_permissions_write_relationships(this.handle, writeBytes);
    }
  }

  /**
   * Permissions service: unary RPCs via FFI, streaming (ReadRelationships, LookupResources, LookupSubjects) via streaming proxy.
   */
  permissions(): v1.ZedClientInterface {
    const h = this.handle;
    const stream = this.streamingClient;

    const unaryCheckPermission = (
      input: v1.CheckPermissionRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.CheckPermissionResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        CheckPermissionRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_permissions_check_permission(h, bytes);
        const resp = CheckPermissionResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const unaryWriteRelationships = (
      input: v1.WriteRelationshipsRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.WriteRelationshipsResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        WriteRelationshipsRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_permissions_write_relationships(h, bytes);
        const resp = WriteRelationshipsResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const unaryDeleteRelationships = (
      input: v1.DeleteRelationshipsRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.DeleteRelationshipsResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        DeleteRelationshipsRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_permissions_delete_relationships(h, bytes);
        const resp = DeleteRelationshipsResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const unaryCheckBulkPermissions = (
      input: v1.CheckBulkPermissionsRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.CheckBulkPermissionsResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        CheckBulkPermissionsRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_permissions_check_bulk_permissions(h, bytes);
        const resp = CheckBulkPermissionsResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const unaryExpandPermissionTree = (
      input: v1.ExpandPermissionTreeRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.ExpandPermissionTreeResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        ExpandPermissionTreeRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_permissions_expand_permission_tree(h, bytes);
        const resp = ExpandPermissionTreeResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const promises = {
      ...stream.promises,
      readRelationships: stream.promises.readRelationships.bind(
        stream.promises
      ),
      lookupResources: stream.promises.lookupResources.bind(stream.promises),
      lookupSubjects: stream.promises.lookupSubjects.bind(stream.promises),
      checkPermission: (input: v1.CheckPermissionRequest) =>
        new Promise<v1.CheckPermissionResponse>((resolve, reject) => {
          unaryCheckPermission(input, undefined, undefined, (err, value) =>
            err ? reject(err) : resolve(value!)
          );
        }),
      writeRelationships: (input: v1.WriteRelationshipsRequest) =>
        new Promise<v1.WriteRelationshipsResponse>((resolve, reject) => {
          unaryWriteRelationships(input, undefined, undefined, (err, value) =>
            err ? reject(err) : resolve(value!)
          );
        }),
      deleteRelationships: (input: v1.DeleteRelationshipsRequest) =>
        new Promise<v1.DeleteRelationshipsResponse>((resolve, reject) => {
          unaryDeleteRelationships(input, undefined, undefined, (err, value) =>
            err ? reject(err) : resolve(value!)
          );
        }),
      checkBulkPermissions: (input: v1.CheckBulkPermissionsRequest) =>
        new Promise<v1.CheckBulkPermissionsResponse>((resolve, reject) => {
          unaryCheckBulkPermissions(
            input,
            undefined,
            undefined,
            (err, value) => (err ? reject(err) : resolve(value!))
          );
        }),
      expandPermissionTree: (input: v1.ExpandPermissionTreeRequest) =>
        new Promise<v1.ExpandPermissionTreeResponse>((resolve, reject) => {
          unaryExpandPermissionTree(
            input,
            undefined,
            undefined,
            (err, value) => (err ? reject(err) : resolve(value!))
          );
        }),
    };

    return new Proxy(stream, {
      get(target, prop) {
        if (prop === "promises") return promises;
        if (prop === "checkPermission") return unaryCheckPermission;
        if (prop === "writeRelationships") return unaryWriteRelationships;
        if (prop === "deleteRelationships") return unaryDeleteRelationships;
        if (prop === "checkBulkPermissions") return unaryCheckBulkPermissions;
        if (prop === "expandPermissionTree") return unaryExpandPermissionTree;
        return (target as Record<string, unknown>)[prop as string];
      },
    }) as unknown as v1.ZedClientInterface;
  }

  /**
   * Schema service (ReadSchema, WriteSchema, etc.) via FFI.
   */
  schema(): v1.ZedClientInterface {
    const h = this.handle;
    const stream = this.streamingClient;

    const unaryReadSchema = (
      input: v1.ReadSchemaRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.ReadSchemaResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        ReadSchemaRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_schema_read_schema(h, bytes);
        const resp = ReadSchemaResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const unaryWriteSchema = (
      input: v1.WriteSchemaRequest,
      metadata?: grpc.Metadata,
      options?: grpc.CallOptions,
      callback?: (
        err: grpc.ServiceError | null,
        value?: v1.WriteSchemaResponse
      ) => void
    ): grpc.ClientUnaryCall => {
      const bytes = new Uint8Array(
        WriteSchemaRequest.toBinary(input) as ArrayLike<number>
      );
      try {
        const out = spicedb_schema_write_schema(h, bytes);
        const resp = WriteSchemaResponse.fromBinary(out);
        if (callback) {
          setImmediate(() => callback(null, resp));
        }
      } catch (err) {
        if (callback) {
          setImmediate(() => callback(err as grpc.ServiceError, undefined));
        }
      }
      return {} as grpc.ClientUnaryCall;
    };

    const promises = {
      ...stream.promises,
      readSchema: (input: v1.ReadSchemaRequest) =>
        new Promise<v1.ReadSchemaResponse>((resolve, reject) => {
          unaryReadSchema(input, undefined, undefined, (err, value) =>
            err ? reject(err) : resolve(value!)
          );
        }),
      writeSchema: (input: v1.WriteSchemaRequest) =>
        new Promise<v1.WriteSchemaResponse>((resolve, reject) => {
          unaryWriteSchema(input, undefined, undefined, (err, value) =>
            err ? reject(err) : resolve(value!)
          );
        }),
    };

    return new Proxy(stream, {
      get(target, prop) {
        if (prop === "promises") return promises;
        if (prop === "readSchema") return unaryReadSchema;
        if (prop === "writeSchema") return unaryWriteSchema;
        return (target as Record<string, unknown>)[prop as string];
      },
    }) as unknown as v1.ZedClientInterface;
  }

  /**
   * Watch service (streaming) via streaming proxy.
   */
  watch(): v1.ZedClientInterface {
    return this.streamingClient;
  }

  /**
   * Streaming address (Unix path or host:port) for streaming APIs.
   */
  streamingAddress(): string {
    return this._streamingAddress;
  }

  /**
   * Dispose the instance and close the streaming channel.
   */
  close(): void {
    if (this._closed) return;
    this._closed = true;
    this.streamingClient.close();
    spicedb_dispose(this.handle);
  }
}
